contracts/FeralfileArtworkV4.sol

```diff
@@ -241,24 +231,24 @@ contract FeralfileExhibitionV4 is
             saleData_.expiryTime > block.timestamp,
             "FeralfileExhibitionV4: sale is expired"
         );
-        require(
-            saleData_.price == msg.value,
-            "FeralfileExhibitionV4: invalid payable amount and price"
-        );
 
-        if (saleData_.payByVaultContract) {
-            // pay eth by vault contract
-            FeralfileVault(payable(_vaultAddress)).pay(saleData_.cost);
-        }
+        saleData_.payByVaultContract
+            ? Vault(payable(vault)).pay(saleData_.price)
+            : require(
+                saleData_.price == msg.value,
+                "FeralfileExhibitionV4: invalid payment amount"
+            );
 
-        bytes32 requestHash = keccak256(abi.encode(saleData_));
+        bytes32 requestHash = keccak256(
+            abi.encode(block.chainid, address(this), saleData_)
+        );
 
         require(
             isValidRequest(requestHash, signer, r_, s_, v_),
             "FeralfileExhibitionV4: invalid signature"
         );
 
-        uint256 itemCost = saleData_.cost / saleData_.tokenIds.length;
+        uint256 itemRevenue = (saleData_.price - saleData_.cost) / saleData_.tokenIds.length;
```

Comment:
I think it is better to use SafeMath to deal with fund spliting

Reply:
no need for safemath after 0.8